package exception

import io.BufferedReaderWriter
import java.io.BufferedReader
import java.io.FileReader

/**
 * В Kotlin нет специального ключевого слово для поддержки try with resources, вместо этого в стандартной библиотеке
 * есть функция [use]
 * Функция use реализована как функция-расширение, которая применяется к закрываемым ресурсам; она
 * принимает аргумент с лямбда-выражением. Функция вызывает лямбда-выражение и гарантирует закрытие ресурса
 * вне зависимости от того, как выполнится лямбда-выражение - успешно или с исключением.
 * Конечно, функция use объявлена как встраиваемая, поэтому она не ухудшает производительность
 */

fun main() {

    BufferedReader(FileReader("path")).use {

    }

}